stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/decorator/internal/AvatarDecorator.kt

```diff
@@ -2,38 +2,77 @@ package io.getstream.chat.android.ui.message.list.adapter.viewholder.decorator.i
+import io.getstream.chat.android.ui.message.list.MessageListView
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.GiphyViewHolder
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.MessageDeletedViewHolder
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.MessagePlainTextViewHolder
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.TextAndAttachmentsViewHolder
 
-internal class AvatarDecorator : BaseDecorator() {
+internal class AvatarDecorator(
+    private val showAvatarPredicate: MessageListView.ShowAvatarPredicate = DefaultShowAvatarPredicate(),
+) : BaseDecorator() {
     override fun decorateTextAndAttachmentsMessage(
         viewHolder: TextAndAttachmentsViewHolder,
         data: MessageListItem.MessageItem,
     ) {
-        setupAvatar(viewHolder.binding.avatarView, data)
+        setupAvatar(getAvatarView(viewHolder, data.isMine), data)
     }
 
     override fun decoratePlainTextMessage(
         viewHolder: MessagePlainTextViewHolder,
         data: MessageListItem.MessageItem,
     ) {
-        setupAvatar(viewHolder.binding.avatarView, data)
+        setupAvatar(getAvatarView(viewHolder, data.isMine), data)
     }
 
     override fun decorateGiphyMessage(
         viewHolder: GiphyViewHolder,
         data: MessageListItem.MessageItem,
     ) = Unit
 
+    override fun decorateDeletedMessage(viewHolder: MessageDeletedViewHolder, data: MessageListItem.MessageItem) {
+        setupAvatar(getAvatarView(viewHolder, data.isMine), data)
+    }
+
     private fun setupAvatar(avatarView: AvatarView, data: MessageListItem.MessageItem) {
-        if (data.isTheirs && data.isTheirs && data.isBottomPosition()) {
-            avatarView.isVisible = true
+        val shouldShow = showAvatarPredicate.shouldShow(data)
+
+        avatarView.isVisible = shouldShow
+
+        if (shouldShow) {
             avatarView.setUserData(data.message.user)
+        }
+    }
+
+    private fun getAvatarView(holder: TextAndAttachmentsViewHolder, isMine: Boolean): AvatarView {
+        return if (isMine) {
+            holder.binding.avatarView.isVisible = false
+            holder.binding.avatarMineView
+        } else {
+            holder.binding.avatarMineView.isVisible = false
+            holder.binding.avatarView
+        }
+    }
+
+    private fun getAvatarView(holder: MessagePlainTextViewHolder, isMine: Boolean): AvatarView {
+        return if (isMine) {
+            holder.binding.avatarView.isVisible = false
+            holder.binding.avatarMineView
+        } else {
+            holder.binding.avatarMineView.isVisible = false
+            holder.binding.avatarView
+        }
+    }
+
+    private fun getAvatarView(holder: MessageDeletedViewHolder, isMine: Boolean): AvatarView {
+        return if (isMine) {
+            holder.binding.avatarView.isVisible = false
```

Comment:
Something that just occurred to me (a bit late, I know): would it make our lives easier or harder if we used the avatars themselves to provide the margin, by just making them `INVISIBLE` instead of `GONE` when they're not needed? In that case, we'd need attributes to toggle all avatars on/off on each side of the message list, I suppose.

Reply:
I thought about it. If we use a flag, we could just set a hard coded margin so the user wouldn't be able to provide a "wrong" margin... But then we would be taking away a big degree of customization, which I consider to be worse.

About the Invisible/Gone. The logic would be complex... For 0 avatars visible we would use Invisible, but for more than one, we would make Gone... We would need to keep track of the list. Set a custom margin is easier in the case

Reply:
Let's check to be sure that something that's not 0 or the default value makes sense (as per my next comment), otherwise we can go with the custom margin approach

Reply:
The avatar is not the one creating the margin anymore, so different values than default or 0 will be needed if the size of avatar changes (if the user want to keep the same margins from the avatar view)